// @nexus-project/adapter-sdk-ts/dist/index.js
import fs from "node:fs";
import { z } from "zod";
import util from "node:util";
import readline from "node:readline";

//#region src/runtime-context.ts
const ADAPTER_CONTEXT_ENV_VAR = "NEXUS_ADAPTER_CONTEXT_PATH";
const ADAPTER_STATE_DIR_ENV_VAR = "NEXUS_ADAPTER_STATE_DIR";
const AdapterRuntimeCredentialSchema = z.object({
	kind: z.string(),
	value: z.string(),
	fields: z.record(z.string(), z.string()).optional(),
	auth_id: z.string().optional(),
	type: z.string().optional(),
	ref: z.string().optional(),
	service: z.string().optional(),
	account: z.string().optional()
}).catchall(z.unknown());
const AdapterRuntimeContextFileSchema = z.object({
	version: z.number().optional(),
	platform: z.string(),
	connection_id: z.string(),
	config: z.record(z.string(), z.unknown()).default({}),
	credential: AdapterRuntimeCredentialSchema.optional()
}).catchall(z.unknown());
function readAdapterRuntimeContextFile(pathValue) {
	const raw = fs.readFileSync(pathValue, "utf8");
	const parsed = JSON.parse(raw);
	const file = AdapterRuntimeContextFileSchema.parse(parsed);
	return {
		platform: file.platform,
		connection_id: file.connection_id,
		config: file.config,
		...file.credential ? { credential: file.credential } : {},
		raw: file
	};
}
function loadAdapterRuntimeContext(env = process.env) {
	const pathValue = env[ADAPTER_CONTEXT_ENV_VAR]?.trim();
	if (!pathValue) return null;
	return readAdapterRuntimeContextFile(pathValue);
}
function requireAdapterRuntimeContext(env = process.env) {
	const context = loadAdapterRuntimeContext(env);
	if (!context) throw new Error(`Missing adapter runtime context (expected $${ADAPTER_CONTEXT_ENV_VAR} to point at runtime-context.json)`);
	return context;
}
function loadAdapterStateDir(env = process.env) {
	const value = env[ADAPTER_STATE_DIR_ENV_VAR]?.trim();
	return value ? value : null;
}
function requireAdapterStateDir(env = process.env) {
	const value = loadAdapterStateDir(env);
	if (!value) throw new Error(`Missing adapter state dir (expected $${ADAPTER_STATE_DIR_ENV_VAR} to point at a writable adapter state directory)`);
	return value;
}

//#endregion
//#region src/define.ts
function method(declaration) {
	return declaration;
}
function defineAdapter(config) {
	const methodDefinitions = config.methods ?? {};
	const operations = {
		"adapter.info": async () => buildAdapterInfo(config, methodDefinitions),
		"adapter.accounts.list": async (ctx) => {
			const definedCtx = await createDefinedContext(config, ctx, ctx.runtime?.connection_id);
			if (config.connection?.accounts) return await config.connection.accounts(definedCtx);
			return defaultAccounts(ctx);
		},
		"adapter.health": async (ctx, args) => {
			const definedCtx = await createDefinedContext(config, ctx, args.connection_id);
			if (config.connection?.health) return {
				...await config.connection.health(definedCtx),
				connection_id: args.connection_id
			};
			return {
				connected: true,
				connection_id: args.connection_id
			};
		},
		methods: buildMethodHandlers(config, methodDefinitions)
	};
	if (config.ingest?.monitor) operations["adapter.monitor.start"] = async (ctx, args, emit) => {
		const definedCtx = await createDefinedContext(config, ctx, args.connection_id);
		await config.ingest.monitor(definedCtx, emit);
	};
	if (config.ingest?.backfill) operations["records.backfill"] = async (ctx, args, emit) => {
		const definedCtx = await createDefinedContext(config, ctx, args.connection_id);
		await config.ingest.backfill(definedCtx, { since: args.since }, emit);
	};
	if (config.delivery?.send) operations["channels.send"] = async (ctx, req) => {
		const definedCtx = await createDefinedContext(config, ctx, req.target.connection_id);
		return await config.delivery.send(definedCtx, req);
	};
	if (config.delivery?.stream) operations["channels.stream"] = config.delivery.stream;
	if (config.setup?.start) operations["adapter.setup.start"] = async (ctx, req) => {
		const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
		return await config.setup.start(definedCtx, req);
	};
	if (config.setup?.submit) operations["adapter.setup.submit"] = async (ctx, req) => {
		const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
		return await config.setup.submit(definedCtx, req);
	};
	if (config.setup?.status) operations["adapter.setup.status"] = async (ctx, req) => {
		const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
		return await config.setup.status(definedCtx, req);
	};
	if (config.setup?.cancel) operations["adapter.setup.cancel"] = async (ctx, req) => {
		const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
		return await config.setup.cancel(definedCtx, req);
	};
	if (config.control) operations["adapter.control.start"] = config.control;
	return { operations };
}
async function createDefinedContext(config, ctx, connectionId) {
	const client = config.client?.create ? await config.client.create({
		ctx,
		connectionId
	}) : void 0;
	return {
		...ctx,
		connectionId,
		client
	};
}
function defaultAccounts(ctx) {
	const connectionId = ctx.runtime?.connection_id?.trim();
	if (!connectionId) return [];
	return [{
		id: connectionId,
		status: "ready",
		...ctx.runtime?.credential?.ref ? { credential_ref: ctx.runtime.credential.ref } : {}
	}];
}
function buildMethodHandlers(config, methods) {
	if (Object.keys(methods).length === 0) return {};
	return Object.fromEntries(Object.entries(methods).map(([name, declaration]) => [name, async (ctx, req) => {
		const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
		return await declaration.handler(definedCtx, req);
	}]));
}
function buildAdapterInfo(config, methods) {
	const operations = [
		"adapter.info",
		"adapter.accounts.list",
		"adapter.health",
		...config.ingest?.monitor ? ["adapter.monitor.start"] : [],
		...config.ingest?.backfill ? ["records.backfill"] : [],
		...config.delivery?.send ? ["channels.send"] : [],
		...config.delivery?.stream ? ["channels.stream"] : [],
		...config.setup?.start ? ["adapter.setup.start"] : [],
		...config.setup?.submit ? ["adapter.setup.submit"] : [],
		...config.setup?.status ? ["adapter.setup.status"] : [],
		...config.setup?.cancel ? ["adapter.setup.cancel"] : [],
		...config.control ? ["adapter.control.start"] : []
	];
	const methodDescriptors = Object.entries(methods).map(([name, declaration]) => buildMethodDescriptor(name, declaration));
	return {
		platform: config.platform,
		name: config.name,
		version: config.version,
		operations,
		methods: methodDescriptors,
		...config.credential_service ? { credential_service: config.credential_service } : {},
		multi_account: config.multi_account ?? false,
		platform_capabilities: config.capabilities,
		...config.auth ? { auth: config.auth } : {},
		...methodDescriptors.length > 0 ? { methodCatalog: config.methodCatalog ?? {
			source: "manifest",
			namespace: config.platform
		} } : config.methodCatalog ? { methodCatalog: config.methodCatalog } : {}
	};
}
function buildMethodDescriptor(name, declaration) {
	return {
		name,
		...declaration.description !== void 0 ? { description: declaration.description } : {},
		...declaration.action ? { action: declaration.action } : {},
		...declaration.params ? { params: declaration.params } : {},
		...declaration.response ? { response: declaration.response } : {},
		surfaces: declaration.surfaces ?? ["ws.control", "http.control"],
		connection_required: declaration.connection_required ?? true,
		mutates_remote: typeof declaration.mutates_remote === "boolean" ? declaration.mutates_remote : declaration.action === "write",
		...declaration.context_hints ? { context_hints: declaration.context_hints } : {}
	};
}

//#endregion
//#region src/credentials.ts
function readEnv(name) {
	const value = process.env[name]?.trim();
	return value ? value : void 0;
}
function readCredential(ctx, options = {}) {
	const fields = options.fields ?? [];
	for (const field of fields) {
		const value = ctx.runtime?.credential?.fields?.[field]?.trim();
		if (value) return value;
	}
	if (options.allowValue !== false) {
		const runtimeValue = ctx.runtime?.credential?.value?.trim();
		if (runtimeValue) return runtimeValue;
	}
	const envVars = options.env ?? [];
	for (const name of envVars) {
		const value = readEnv(name);
		if (value) return value;
	}
}
function requireCredential(ctx, options = {}) {
	const value = readCredential(ctx, options);
	if (value) return value;
	const label = options.label?.trim() || "credential";
	const sources = [];
	if ((options.fields ?? []).length > 0) sources.push(`runtime credential fields ${JSON.stringify(options.fields)}`);
	if (options.allowValue !== false) sources.push("runtime credential value");
	if ((options.env ?? []).length > 0) sources.push(`environment ${JSON.stringify(options.env)}`);
	const suffix = sources.length > 0 ? ` (looked in ${sources.join(", ")})` : "";
	throw new Error(`missing ${label}${suffix}`);
}

//#endregion
//#region src/targets.ts
function asNonEmptyString(value) {
	const trimmed = value?.trim();
	return trimmed ? trimmed : void 0;
}
function requireContainerTarget(target) {
	const value = asNonEmptyString(target.channel.container_id);
	if (!value) throw new Error("target.channel.container_id is required");
	return value;
}
function readThreadTarget(target) {
	return asNonEmptyString(target.channel.thread_id);
}
function readReplyToTarget(target) {
	return asNonEmptyString(target.reply_to_id);
}

//#endregion
//#region src/retry.ts
async function sleepWithSignal(signal, ms) {
	if (ms <= 0 || signal.aborted) return;
	await new Promise((resolve) => {
		const timer = setTimeout(() => {
			signal.removeEventListener("abort", onAbort);
			resolve();
		}, ms);
		const onAbort = () => {
			clearTimeout(timer);
			signal.removeEventListener("abort", onAbort);
			resolve();
		};
		signal.addEventListener("abort", onAbort, { once: true });
	});
}
function parseRetryAfterMs(value) {
	if (typeof value === "number" && Number.isFinite(value) && value >= 0) return Math.trunc(value);
	if (typeof value !== "string") return;
	const trimmed = value.trim();
	if (!trimmed) return;
	const seconds = Number(trimmed);
	if (Number.isFinite(seconds) && seconds >= 0) return Math.trunc(seconds * 1e3);
	const dateMs = Date.parse(trimmed);
	if (Number.isNaN(dateMs)) return;
	return Math.max(0, dateMs - Date.now());
}
async function withRetry(work, options = {}) {
	const attempts = Math.max(1, Math.trunc(options.attempts ?? 3));
	const factor = options.factor && options.factor > 0 ? options.factor : 2;
	const maxDelayMs = Math.max(0, Math.trunc(options.maxDelayMs ?? 3e4));
	let delayMs = Math.max(0, Math.trunc(options.initialDelayMs ?? 1e3));
	let attempt = 0;
	while (true) {
		attempt += 1;
		try {
			return await work();
		} catch (error) {
			if (!(options.shouldRetry ? options.shouldRetry(error, attempt) : attempt < attempts) || attempt >= attempts || options.signal?.aborted) throw error;
			if (options.onRetry) await options.onRetry(error, attempt, delayMs);
			if (options.signal) await sleepWithSignal(options.signal, delayMs);
			else if (delayMs > 0) await new Promise((resolve) => setTimeout(resolve, delayMs));
			delayMs = Math.min(maxDelayMs, Math.max(delayMs, 1) * factor);
		}
	}
}

//#endregion
//#region src/logger.ts
function createAdapterLogger(opts) {
	const stderr = opts.stderr ?? process.stderr;
	const verbose = opts.verbose;
	const write = (prefix, message, ...args) => {
		const rendered = args.length > 0 ? util.format(message, ...args) : message;
		stderr.write(`${prefix}${rendered}\n`);
	};
	return {
		error: (message, ...args) => write("[ERROR] ", message, ...args),
		info: (message, ...args) => write("[INFO] ", message, ...args),
		debug: (message, ...args) => {
			if (!verbose) return;
			write("[DEBUG] ", message, ...args);
		}
	};
}
function patchConsoleToStderr(log) {
	console.log = (...args) => log.info("%s", util.format(...args));
	console.info = (...args) => log.info("%s", util.format(...args));
	console.warn = (...args) => log.info("%s", util.format(...args));
	console.error = (...args) => log.error("%s", util.format(...args));
}

//#endregion
//#region src/io.ts
function writeJSONLine(stdout, payload) {
	stdout.write(`${JSON.stringify(payload)}\n`);
}
async function* readJSONLines(stream) {
	const rl = readline.createInterface({
		input: stream,
		crlfDelay: Number.POSITIVE_INFINITY
	});
	for await (const line of rl) {
		const trimmed = String(line).trim();
		if (!trimmed) continue;
		yield JSON.parse(trimmed);
	}
}

//#endregion
//#region src/protocol.ts
const AdapterOperationSchema = z.enum([
	"adapter.info",
	"adapter.health",
	"adapter.accounts.list",
	"adapter.monitor.start",
	"adapter.control.start",
	"adapter.setup.start",
	"adapter.setup.submit",
	"adapter.setup.status",
	"adapter.setup.cancel",
	"records.backfill",
	"channels.send",
	"channels.stream",
	"channels.react",
	"channels.edit",
	"channels.delete"
]);
const ChannelCapabilitiesSchema = z.object({
	text_limit: z.number().int().nonnegative().optional(),
	caption_limit: z.number().int().nonnegative().optional(),
	supports_markdown: z.boolean().optional(),
	markdown_flavor: z.string().optional(),
	supports_tables: z.boolean().optional(),
	supports_code_blocks: z.boolean().optional(),
	supports_embeds: z.boolean().optional(),
	supports_threads: z.boolean().optional(),
	supports_reactions: z.boolean().optional(),
	supports_polls: z.boolean().optional(),
	supports_buttons: z.boolean().optional(),
	supports_edit: z.boolean().optional(),
	supports_delete: z.boolean().optional(),
	supports_media: z.boolean().optional(),
	supports_voice_notes: z.boolean().optional(),
	supports_streaming_edit: z.boolean().optional(),
	supports_ptt: z.boolean().optional(),
	supports_streaming: z.boolean().optional(),
	max_message_length: z.number().int().positive().optional(),
	max_attachments: z.number().int().positive().optional()
}).catchall(z.unknown());
const AdapterAuthFieldOptionSchema = z.object({
	label: z.string(),
	value: z.string()
});
const AdapterAuthFieldSchema = z.object({
	name: z.string(),
	label: z.string(),
	type: z.enum([
		"secret",
		"text",
		"select"
	]),
	required: z.boolean(),
	placeholder: z.string().optional(),
	options: z.array(AdapterAuthFieldOptionSchema).optional()
});
const AdapterAuthMethodOAuthSchema = z.object({
	id: z.string(),
	type: z.literal("oauth2"),
	label: z.string(),
	icon: z.string(),
	service: z.string(),
	scopes: z.array(z.string()),
	platformCredentials: z.boolean().optional(),
	platformCredentialUrl: z.string().optional()
});
const AdapterAuthMethodApiKeySchema = z.object({
	id: z.string(),
	type: z.literal("api_key"),
	label: z.string(),
	icon: z.string(),
	service: z.string(),
	fields: z.array(AdapterAuthFieldSchema)
});
const AdapterAuthMethodFileUploadSchema = z.object({
	id: z.string(),
	type: z.literal("file_upload"),
	label: z.string(),
	icon: z.string(),
	accept: z.array(z.string()),
	templateUrl: z.string().optional(),
	maxSize: z.number().int().positive().optional()
});
const AdapterAuthMethodCustomFlowSchema = z.object({
	id: z.string(),
	type: z.literal("custom_flow"),
	label: z.string(),
	icon: z.string(),
	service: z.string(),
	fields: z.array(AdapterAuthFieldSchema).optional()
});
const AdapterAuthMethodSchema = z.discriminatedUnion("type", [
	AdapterAuthMethodOAuthSchema,
	AdapterAuthMethodApiKeySchema,
	AdapterAuthMethodFileUploadSchema,
	AdapterAuthMethodCustomFlowSchema
]);
const AdapterAuthManifestSchema = z.object({
	methods: z.array(AdapterAuthMethodSchema),
	setupGuide: z.string().optional()
});
const AdapterMethodContextHintValueSchema = z.object({
	value: z.unknown(),
	source: z.string(),
	confidence: z.enum([
		"exact",
		"derived",
		"weak"
	])
});
const AdapterMethodContextHintsSchema = z.object({ params: z.record(z.string(), AdapterMethodContextHintValueSchema) });
const AdapterMethodOriginSchema = z.object({
	kind: z.enum([
		"core",
		"app",
		"adapter"
	]),
	package_id: z.string().nullable(),
	package_version: z.string().nullable(),
	declaration_mode: z.enum([
		"manifest",
		"openapi",
		"builtin"
	]),
	declaration_source: z.string(),
	namespace: z.string()
});
const AdapterMethodSchema = z.object({
	name: z.string(),
	description: z.string().nullable().optional(),
	action: z.enum(["read", "write"]).optional(),
	params: z.record(z.string(), z.unknown()).nullable().optional(),
	response: z.record(z.string(), z.unknown()).nullable().optional(),
	surfaces: z.array(z.enum(["ws.control", "http.control"])).optional(),
	connection_required: z.boolean().optional(),
	mutates_remote: z.boolean().optional(),
	context_hints: AdapterMethodContextHintsSchema.optional(),
	origin: AdapterMethodOriginSchema.optional()
});
const AdapterMethodCatalogSchema = z.object({
	source: z.enum(["manifest", "openapi"]).optional(),
	document: z.string().optional(),
	namespace: z.string().optional()
});
const AdapterInfoSchema = z.object({
	platform: z.string(),
	name: z.string(),
	version: z.string(),
	operations: z.array(AdapterOperationSchema),
	methods: z.array(AdapterMethodSchema),
	credential_service: z.string().optional(),
	multi_account: z.boolean(),
	platform_capabilities: ChannelCapabilitiesSchema,
	auth: AdapterAuthManifestSchema.optional(),
	methodCatalog: AdapterMethodCatalogSchema.optional()
});
const AttachmentSchema = z.object({
	id: z.string(),
	filename: z.string().optional(),
	mime_type: z.string(),
	media_type: z.string().optional(),
	size: z.number().int().nonnegative().optional(),
	url: z.string().optional(),
	local_path: z.string().optional(),
	content_hash: z.string().optional(),
	metadata: z.record(z.string(), z.unknown()).optional()
});
const RecipientSchema = z.object({
	id: z.string(),
	name: z.string().optional(),
	avatar_url: z.string().optional()
});
const ContentTypeSchema = z.enum([
	"text",
	"reaction",
	"membership"
]);
const ContainerKindSchema = z.enum(["direct", "group"]);
const AdapterInboundRoutingSchema = z.object({
	adapter: z.string().optional(),
	platform: z.string(),
	connection_id: z.string(),
	sender_id: z.string(),
	sender_name: z.string().optional(),
	receiver_id: z.string().optional(),
	receiver_name: z.string().optional(),
	space_id: z.string().optional(),
	space_name: z.string().optional(),
	container_kind: ContainerKindSchema,
	container_id: z.string(),
	container_name: z.string().optional(),
	thread_id: z.string().optional(),
	thread_name: z.string().optional(),
	reply_to_id: z.string().optional(),
	metadata: z.record(z.string(), z.unknown()).optional()
});
const AdapterInboundPayloadSchema = z.object({
	external_record_id: z.string(),
	timestamp: z.number().int(),
	content: z.string(),
	content_type: ContentTypeSchema,
	reply_to_id: z.string().optional(),
	attachments: z.array(AttachmentSchema).optional(),
	recipients: z.array(z.string()).optional(),
	metadata: z.record(z.string(), z.unknown()).optional()
});
const AdapterInboundRecordSchema = z.object({
	operation: z.literal("record.ingest"),
	routing: AdapterInboundRoutingSchema,
	payload: AdapterInboundPayloadSchema
});
const DeliveryErrorTypeSchema = z.enum([
	"rate_limited",
	"permission_denied",
	"not_found",
	"content_rejected",
	"network",
	"unknown"
]);
const DeliveryErrorSchema = z.object({
	type: DeliveryErrorTypeSchema.optional(),
	message: z.string(),
	retry: z.boolean().optional(),
	retry_after_ms: z.number().int().nonnegative().optional(),
	details: z.record(z.string(), z.unknown()).optional()
});
const DeliveryResultSchema = z.object({
	success: z.boolean(),
	message_ids: z.array(z.string()),
	chunks_sent: z.number().int().nonnegative(),
	total_chars: z.number().int().nonnegative().optional(),
	error: z.union([z.string(), DeliveryErrorSchema]).optional()
});
const ChannelRefSchema = z.object({
	platform: z.string(),
	space_id: z.string().optional(),
	container_kind: ContainerKindSchema.optional(),
	container_id: z.string().optional(),
	thread_id: z.string().optional()
});
const DeliveryTargetSchema = z.object({
	connection_id: z.string(),
	channel: ChannelRefSchema,
	reply_to_id: z.string().optional()
});
const SendRequestSchema = z.object({
	target: DeliveryTargetSchema,
	text: z.string().optional(),
	media: z.string().optional(),
	caption: z.string().optional()
});
const AdapterHealthSchema = z.object({
	connected: z.boolean(),
	connection_id: z.string(),
	last_event_at: z.number().int().optional(),
	error: z.string().optional(),
	details: z.record(z.string(), z.unknown()).optional()
});
const AdapterAccountSchema = z.object({
	id: z.string(),
	display_name: z.string().optional(),
	credential_ref: z.string().optional(),
	status: z.enum([
		"ready",
		"active",
		"error"
	])
});
const AdapterControlEndpointSchema = z.object({
	endpoint_id: z.string(),
	display_name: z.string().optional(),
	platform: z.string().optional(),
	caps: z.array(z.string()),
	commands: z.array(z.string()),
	permissions: z.record(z.string(), z.boolean()).optional()
});
const AdapterControlInvokeErrorSchema = z.object({
	code: z.string().optional(),
	message: z.string().optional()
});
const AdapterControlInvokeRequestFrameSchema = z.object({
	type: z.literal("invoke.request"),
	request_id: z.string(),
	endpoint_id: z.string(),
	command: z.string(),
	payload: z.unknown().optional(),
	timeout_ms: z.number().int().nonnegative().optional(),
	idempotency_key: z.string().optional()
});
const AdapterControlInvokeCancelFrameSchema = z.object({
	type: z.literal("invoke.cancel"),
	request_id: z.string()
});
const AdapterControlInputFrameSchema = z.discriminatedUnion("type", [AdapterControlInvokeRequestFrameSchema, AdapterControlInvokeCancelFrameSchema]);
const AdapterControlEndpointUpsertFrameSchema = z.object({
	type: z.literal("endpoint.upsert"),
	endpoint_id: z.string(),
	display_name: z.string().optional(),
	platform: z.string().optional(),
	caps: z.array(z.string()),
	commands: z.array(z.string()),
	permissions: z.record(z.string(), z.boolean()).optional()
});
const AdapterControlEndpointRemoveFrameSchema = z.object({
	type: z.literal("endpoint.remove"),
	endpoint_id: z.string()
});
const AdapterControlInvokeResultFrameSchema = z.object({
	type: z.literal("invoke.result"),
	request_id: z.string(),
	ok: z.boolean(),
	payload: z.unknown().optional(),
	error: z.union([z.string(), AdapterControlInvokeErrorSchema]).optional()
});
const AdapterControlRecordIngestFrameSchema = z.object({
	type: z.literal("record.ingest"),
	record: z.record(z.string(), z.unknown())
});
const AdapterControlOutputFrameSchema = z.discriminatedUnion("type", [
	AdapterControlEndpointUpsertFrameSchema,
	AdapterControlEndpointRemoveFrameSchema,
	AdapterControlInvokeResultFrameSchema,
	AdapterControlRecordIngestFrameSchema
]);
const StreamEventSchema = z.discriminatedUnion("type", [
	z.object({
		type: z.literal("stream_start"),
		runId: z.string(),
		session_id: z.string().optional(),
		target: DeliveryTargetSchema
	}),
	z.object({
		type: z.literal("token"),
		text: z.string()
	}),
	z.object({
		type: z.literal("tool_status"),
		toolName: z.string(),
		toolCallId: z.string(),
		status: z.enum([
			"started",
			"completed",
			"failed"
		]),
		summary: z.string().optional()
	}),
	z.object({
		type: z.literal("reasoning"),
		text: z.string()
	}),
	z.object({
		type: z.literal("stream_end"),
		runId: z.string(),
		final: z.boolean().optional()
	}),
	z.object({
		type: z.literal("stream_error"),
		error: z.string(),
		partial: z.boolean().optional()
	})
]);
const AdapterStreamStatusSchema = z.discriminatedUnion("type", [
	z.object({
		type: z.literal("message_created"),
		messageId: z.string()
	}),
	z.object({
		type: z.literal("message_updated"),
		messageId: z.string(),
		chars: z.number().int().nonnegative()
	}),
	z.object({
		type: z.literal("message_sent"),
		messageId: z.string(),
		final: z.boolean()
	}),
	z.object({
		type: z.literal("delivery_complete"),
		messageIds: z.array(z.string())
	}),
	z.object({
		type: z.literal("delivery_error"),
		error: z.string()
	}),
	z.object({
		type: z.literal("error"),
		error: z.string()
	})
]);
const AdapterSetupStatusSchema = z.enum([
	"pending",
	"requires_input",
	"completed",
	"failed",
	"cancelled"
]);
const AdapterSetupResultSchema = z.object({
	status: AdapterSetupStatusSchema,
	session_id: z.string().optional(),
	connection_id: z.string().optional(),
	service: z.string().optional(),
	message: z.string().optional(),
	instructions: z.string().optional(),
	fields: z.array(AdapterAuthFieldSchema).optional(),
	secret_fields: z.record(z.string(), z.string()).optional(),
	metadata: z.record(z.string(), z.unknown()).optional()
}).catchall(z.unknown());

//#endregion
//#region src/control.ts
var AdapterControlEndpointRegistry = class {
	endpoints = /* @__PURE__ */ new Map();
	session;
	constructor(session) {
		this.session = session;
	}
	async upsert(endpoint) {
		const parsed = AdapterControlEndpointSchema.parse(endpoint);
		this.endpoints.set(parsed.endpoint_id, { ...parsed });
		await this.session.upsertEndpoint(parsed);
	}
	async remove(endpointId) {
		const parsed = endpointId.trim();
		if (!parsed) throw new Error("endpointId is required");
		this.endpoints.delete(parsed);
		await this.session.removeEndpoint(parsed);
	}
	get(endpointId) {
		const parsed = endpointId.trim();
		if (!parsed) return;
		const value = this.endpoints.get(parsed);
		return value ? { ...value } : void 0;
	}
	list() {
		return [...this.endpoints.values()].map((value) => ({ ...value })).toSorted((a, b) => a.endpoint_id.localeCompare(b.endpoint_id));
	}
};
var AdapterControlSession = class {
	stdin;
	stdout;
	signal;
	validateFrames;
	log;
	constructor(opts) {
		this.stdin = opts.stdin;
		this.stdout = opts.stdout;
		this.signal = opts.signal;
		this.validateFrames = opts.validateFrames ?? true;
		this.log = opts.log;
	}
	createEndpointRegistry() {
		return new AdapterControlEndpointRegistry(this);
	}
	async send(frame) {
		const payload = this.validateFrames ? AdapterControlOutputFrameSchema.parse(frame) : frame;
		writeJSONLine(this.stdout, payload);
	}
	async upsertEndpoint(endpoint) {
		await this.send({
			type: "endpoint.upsert",
			endpoint_id: endpoint.endpoint_id,
			display_name: endpoint.display_name,
			platform: endpoint.platform,
			caps: endpoint.caps,
			commands: endpoint.commands,
			permissions: endpoint.permissions
		});
	}
	async removeEndpoint(endpointId) {
		const parsed = endpointId.trim();
		if (!parsed) throw new Error("endpointId is required");
		await this.send({
			type: "endpoint.remove",
			endpoint_id: parsed
		});
	}
	async emitRecordIngest(record) {
		const frame = {
			type: "record.ingest",
			record
		};
		await this.send(frame);
	}
	async respondInvokeResult(params) {
		const requestID = params.request_id.trim();
		if (!requestID) throw new Error("request_id is required");
		const frame = {
			type: "invoke.result",
			request_id: requestID,
			ok: params.ok === true,
			payload: params.payload,
			error: normalizeInvokeError(params.error)
		};
		await this.send(frame);
	}
	async serve(handlers) {
		for await (const raw of readJSONLines(this.stdin)) {
			if (this.signal?.aborted) return;
			const frame = this.parseInputFrame(raw);
			if (!frame) continue;
			if (frame.type === "invoke.cancel") {
				await handlers.onCancel?.(frame);
				continue;
			}
			if (frame.type === "invoke.request") {
				if (!handlers.onInvoke) continue;
				try {
					const result = await handlers.onInvoke(frame);
					if (result === void 0) continue;
					await this.respondInvokeResult({
						request_id: frame.request_id,
						ok: result.ok,
						payload: result.payload,
						error: result.error
					});
				} catch (error) {
					await this.respondInvokeResult({
						request_id: frame.request_id,
						ok: false,
						error: { message: errorToString$2(error) }
					});
				}
			}
		}
	}
	parseInputFrame(raw) {
		if (!this.validateFrames) {
			const candidate = raw;
			if (candidate && typeof candidate === "object" && "type" in candidate && (candidate.type === "invoke.request" || candidate.type === "invoke.cancel")) return candidate;
			logControlWarning(this.log, "control session ignored malformed frame");
			return null;
		}
		try {
			return AdapterControlInputFrameSchema.parse(raw);
		} catch (error) {
			logControlWarning(this.log, `control session ignored malformed frame: ${errorToString$2(error)}`);
			return null;
		}
	}
};
function createAdapterControlSession(opts) {
	return new AdapterControlSession(opts);
}
function normalizeInvokeError(value) {
	if (value === null || value === void 0) return;
	if (typeof value === "string") return value;
	return AdapterControlInvokeErrorSchema.parse(value);
}
function errorToString$2(err) {
	return err instanceof Error ? err.message : String(err);
}
function logControlWarning(log, message) {
	if (!log) return;
	if (typeof log.warn === "function") {
		log.warn(message);
		return;
	}
	log.info?.(message);
}

//#endregion
//#region src/stream.ts
function emitStreamStatus(status, stdout = process.stdout, validate = true) {
	writeJSONLine(stdout, validate ? AdapterStreamStatusSchema.parse(status) : status);
}
async function handleStream(opts) {
	const stdin = opts.stdin ?? process.stdin;
	const stdout = opts.stdout ?? process.stdout;
	const validate = opts.validate ?? true;
	const ctx = opts.ctx;
	const handlers = opts.handlers;
	const dispatchError = async (log, message) => {
		log.error("%s", message);
		emitStreamStatus({
			type: "delivery_error",
			error: message
		}, stdout, validate);
	};
	for await (const raw of readJSONLines(stdin)) {
		if (ctx.signal.aborted) {
			ctx.log.info("stream shutting down (signal aborted)");
			return;
		}
		let event;
		try {
			event = validate ? StreamEventSchema.parse(raw) : raw;
		} catch (err) {
			await dispatchError(ctx.log, `stream: failed to parse event: ${errorToString$1(err)}`);
			continue;
		}
		try {
			switch (event.type) {
				case "stream_start":
					if (handlers.onStart) await handlers.onStart(ctx, event);
					break;
				case "token":
					if (handlers.onToken) await handlers.onToken(ctx, event);
					break;
				case "tool_status":
					if (handlers.onToolStatus) await handlers.onToolStatus(ctx, event);
					break;
				case "reasoning":
					if (handlers.onReasoning) await handlers.onReasoning(ctx, event);
					break;
				case "stream_end":
					if (handlers.onEnd) await handlers.onEnd(ctx, event);
					break;
				case "stream_error":
					if (handlers.onError) await handlers.onError(ctx, event);
					break;
				default: ctx.log.debug("stream: unknown event type: %s", event.type);
			}
		} catch (err) {
			await dispatchError(ctx.log, `stream handler error for ${event.type}: ${errorToString$1(err)}`);
		}
	}
}
function errorToString$1(err) {
	return err instanceof Error ? err.message : String(err);
}

//#endregion
//#region src/run.ts
async function runAdapter(adapter, opts = {}) {
	const argv = opts.argv ?? process.argv;
	const env = opts.env ?? process.env;
	const stdin = opts.stdin ?? process.stdin;
	const stdout = opts.stdout ?? process.stdout;
	const stderr = opts.stderr ?? process.stderr;
	const requireRuntimeContext = opts.requireRuntimeContext ?? true;
	const validateOutput = opts.validateOutput ?? true;
	const doPatchConsole = opts.patchConsole ?? true;
	const installSignalHandlers = opts.installSignalHandlers ?? true;
	const command = String(argv[2] ?? "").trim();
	const args = argv.slice(3);
	const log = createAdapterLogger({
		verbose: args.includes("--verbose") || args.includes("-v"),
		stderr
	});
	if (doPatchConsole) patchConsoleToStderr(log);
	if (!command || command === "help" || command === "--help" || command === "-h") {
		printUsage(argv[1] ?? "adapter", stderr);
		return command ? 0 : 1;
	}
	const filteredArgs = args.filter((a) => a !== "--verbose" && a !== "-v");
	const controller = new AbortController();
	if (installSignalHandlers) {
		const onSignal = () => controller.abort();
		process.on("SIGINT", onSignal);
		process.on("SIGTERM", onSignal);
	}
	let runtime = null;
	if (command !== "adapter.info" && command !== "adapter.setup.start" && command !== "adapter.setup.submit" && command !== "adapter.setup.status" && command !== "adapter.setup.cancel") try {
		runtime = requireRuntimeContext ? requireAdapterRuntimeContext(env) : loadAdapterRuntimeContext(env);
	} catch (err) {
		log.error("%s", errorToString(err));
		return 1;
	}
	const ctx = {
		signal: controller.signal,
		runtime,
		log,
		stdout,
		stderr
	};
	const operations = adapter.operations;
	try {
		switch (command) {
			case "adapter.info": {
				const handler = operations["adapter.info"];
				if (!handler) throw new Error("adapter.info not supported by this adapter");
				const info = await handler(ctx);
				writeJSONLine(stdout, validateOutput ? AdapterInfoSchema.parse(info) : info);
				return 0;
			}
			case "adapter.monitor.start": {
				const handler = operations["adapter.monitor.start"];
				if (!handler) throw new Error("adapter.monitor.start not supported by this adapter");
				const connectionID = requireFlag(filteredArgs, "--connection");
				const emit = (record) => {
					writeJSONLine(stdout, validateOutput ? AdapterInboundRecordSchema.parse(record) : record);
				};
				ctx.log.info("monitor starting for connection %s", JSON.stringify(connectionID));
				await handler(ctx, { connection_id: connectionID }, emit);
				ctx.log.info("monitor stopped cleanly");
				return 0;
			}
			case "records.backfill": {
				const handler = operations["records.backfill"];
				if (!handler) throw new Error("records.backfill not supported by this adapter");
				const connectionID = requireFlag(filteredArgs, "--connection");
				const since = parseDate(requireFlag(filteredArgs, "--since"));
				const emit = (record) => {
					writeJSONLine(stdout, validateOutput ? AdapterInboundRecordSchema.parse(record) : record);
				};
				ctx.log.info("backfill starting for connection %s since %s", JSON.stringify(connectionID), since.toISOString());
				await handler(ctx, {
					connection_id: connectionID,
					since
				}, emit);
				ctx.log.info("backfill completed");
				return 0;
			}
			case "channels.send": {
				const handler = operations["channels.send"];
				if (!handler) throw new Error("channels.send not supported by this adapter");
				const connectionID = requireFlag(filteredArgs, "--connection");
				const targetRaw = requireFlag(filteredArgs, "--target-json");
				const target = DeliveryTargetSchema.parse(parseJsonObjectFlag(targetRaw, "--target-json"));
				if (target.connection_id !== connectionID) throw new Error(`--connection ${JSON.stringify(connectionID)} does not match target.connection_id ${JSON.stringify(target.connection_id)}`);
				const text = readFlag(filteredArgs, "--text");
				const media = readFlag(filteredArgs, "--media");
				const caption = readFlag(filteredArgs, "--caption");
				if (!text && !media) throw new Error("channels.send requires --text or --media");
				if (text && media) throw new Error("channels.send must not specify both --text and --media");
				if (caption && !media) throw new Error("channels.send --caption requires --media");
				const req = {
					target,
					...text ? { text } : {},
					...media ? { media } : {},
					...caption ? { caption } : {}
				};
				const parsedReq = validateOutput ? SendRequestSchema.parse(req) : req;
				try {
					const result = await handler(ctx, parsedReq);
					writeJSONLine(stdout, validateOutput ? DeliveryResultSchema.parse(result) : result);
					return 0;
				} catch (err) {
					const failure = {
						success: false,
						message_ids: [],
						chunks_sent: 0,
						error: {
							type: "unknown",
							message: errorToString(err),
							retry: false
						}
					};
					writeJSONLine(stdout, validateOutput ? DeliveryResultSchema.parse(failure) : failure);
					return 0;
				}
			}
			case "adapter.health": {
				const handler = operations["adapter.health"];
				if (!handler) throw new Error("adapter.health not supported by this adapter");
				const connectionID = requireFlag(filteredArgs, "--connection");
				try {
					const health = await handler(ctx, { connection_id: connectionID });
					writeJSONLine(stdout, validateOutput ? AdapterHealthSchema.parse(health) : health);
					return 0;
				} catch (err) {
					const health = {
						connected: false,
						connection_id: connectionID,
						error: errorToString(err)
					};
					writeJSONLine(stdout, validateOutput ? AdapterHealthSchema.parse(health) : health);
					return 0;
				}
			}
			case "adapter.accounts.list": {
				const handler = operations["adapter.accounts.list"];
				if (!handler) throw new Error("adapter.accounts.list not supported by this adapter");
				const accounts = await handler(ctx);
				writeJSONLine(stdout, validateOutput ? zArray(AdapterAccountSchema).parse(accounts) : accounts);
				return 0;
			}
			case "adapter.setup.start": {
				const handler = operations["adapter.setup.start"];
				if (!handler) throw new Error("adapter.setup.start not supported by this adapter");
				const result = await handler(ctx, readAdapterSetupRequest(filteredArgs, false));
				writeJSONLine(stdout, validateOutput ? AdapterSetupResultSchema.parse(result) : result);
				return 0;
			}
			case "adapter.setup.submit": {
				const handler = operations["adapter.setup.submit"];
				if (!handler) throw new Error("adapter.setup.submit not supported by this adapter");
				const result = await handler(ctx, readAdapterSetupRequest(filteredArgs, true));
				writeJSONLine(stdout, validateOutput ? AdapterSetupResultSchema.parse(result) : result);
				return 0;
			}
			case "adapter.setup.status": {
				const handler = operations["adapter.setup.status"];
				if (!handler) throw new Error("adapter.setup.status not supported by this adapter");
				const result = await handler(ctx, readAdapterSetupRequest(filteredArgs, true));
				writeJSONLine(stdout, validateOutput ? AdapterSetupResultSchema.parse(result) : result);
				return 0;
			}
			case "adapter.setup.cancel": {
				const handler = operations["adapter.setup.cancel"];
				if (!handler) throw new Error("adapter.setup.cancel not supported by this adapter");
				const result = await handler(ctx, readAdapterSetupRequest(filteredArgs, true));
				writeJSONLine(stdout, validateOutput ? AdapterSetupResultSchema.parse(result) : result);
				return 0;
			}
			case "adapter.control.start": {
				const handler = operations["adapter.control.start"];
				if (!handler) throw new Error("adapter.control.start not supported by this adapter");
				const connectionID = requireFlag(filteredArgs, "--connection");
				const session = createAdapterControlSession({
					stdin,
					stdout,
					signal: ctx.signal,
					validateFrames: validateOutput,
					log: ctx.log
				});
				await handler(ctx, { connection_id: connectionID }, session);
				return 0;
			}
			case "channels.stream": {
				const handler = operations["channels.stream"];
				if (!handler) throw new Error("channels.stream not supported by this adapter");
				requireFlag(filteredArgs, "--connection");
				ctx.log.info("stream handler starting");
				await handleStream({
					ctx,
					handlers: handler,
					stdin,
					stdout,
					validate: validateOutput
				});
				ctx.log.info("stream handler stopped cleanly");
				return 0;
			}
			default: {
				const handler = operations.methods?.[command];
				if (!handler) {
					stderr.write(`unknown command: ${command}\n`);
					printUsage(argv[1] ?? "adapter", stderr);
					return 1;
				}
				const connectionID = readFlag(filteredArgs, "--connection");
				const payloadRaw = readFlag(filteredArgs, "--payload-json");
				const payload = payloadRaw ? parseJsonObjectFlag(payloadRaw, "--payload-json") : void 0;
				writeJSONLine(stdout, await handler(ctx, {
					...connectionID ? { connection_id: connectionID } : {},
					...payload ? { payload } : {}
				}) ?? {});
				return 0;
			}
		}
	} catch (err) {
		log.error("%s", errorToString(err));
		return 1;
	}
}
function printUsage(name, stderr) {
	stderr.write(`Usage: ${name} <operation> [flags]\n\n`);
	stderr.write("Operations:\n");
	stderr.write("  adapter.info\n");
	stderr.write("  adapter.monitor.start --connection <id>\n");
	stderr.write("  records.backfill --connection <id> --since <date>\n");
	stderr.write("  channels.send --connection <id> --target-json <json> --text \"...\"\n");
	stderr.write("  adapter.health --connection <id>\n");
	stderr.write("  adapter.accounts.list\n");
	stderr.write("  adapter.setup.start [--connection <id>] [--session-id <id>] [--payload-json <json>]\n");
	stderr.write("  adapter.setup.submit --session-id <id> [--connection <id>] [--payload-json <json>]\n");
	stderr.write("  adapter.setup.status --session-id <id> [--connection <id>]\n");
	stderr.write("  adapter.setup.cancel --session-id <id> [--connection <id>]\n");
	stderr.write("  adapter.control.start --connection <id>\n");
	stderr.write("  channels.stream --connection <id>\n\n");
	stderr.write("  <namespace.method> [--connection <id>] [--payload-json <json>]\n\n");
	stderr.write("Global flags:\n");
	stderr.write("  --verbose, -v                     Enable debug logging\n");
}
function readFlag(args, name) {
	const eqPrefix = `${name}=`;
	for (let i = 0; i < args.length; i++) {
		const a = args[i];
		if (a === name) {
			const v = args[i + 1];
			return typeof v === "string" ? v : void 0;
		}
		if (a.startsWith(eqPrefix)) return a.slice(eqPrefix.length);
	}
}
function requireFlag(args, name) {
	const v = readFlag(args, name);
	if (!v || !v.trim()) throw new Error(`Missing required flag: ${name}`);
	return v;
}
function parseJsonObjectFlag(raw, flagName) {
	let parsed;
	try {
		parsed = JSON.parse(raw);
	} catch (error) {
		throw new Error(`${flagName} must be valid JSON: ${errorToString(error)}`);
	}
	if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error(`${flagName} must decode to a JSON object`);
	return parsed;
}
function readAdapterSetupRequest(args, requireSessionID) {
	const connectionID = readFlag(args, "--connection");
	const sessionID = requireSessionID ? requireFlag(args, "--session-id") : readFlag(args, "--session-id");
	const payloadRaw = readFlag(args, "--payload-json");
	const payload = payloadRaw ? parseJsonObjectFlag(payloadRaw, "--payload-json") : void 0;
	return {
		...connectionID ? { connection_id: connectionID } : {},
		...sessionID ? { session_id: sessionID } : {},
		...payload ? { payload } : {}
	};
}
function parseDate(s) {
	const trimmed = s.trim();
	if (!trimmed) throw new Error("date is required");
	const parsed = Date.parse(trimmed);
	if (!Number.isNaN(parsed)) return new Date(parsed);
	const m = /^(\d{4})-(\d{2})-(\d{2})$/u.exec(trimmed);
	if (m) return /* @__PURE__ */ new Date(`${m[1]}-${m[2]}-${m[3]}T00:00:00Z`);
	throw new Error("unrecognized date format (expected ISO 8601 or YYYY-MM-DD)");
}
function errorToString(err) {
	return err instanceof Error ? err.message : String(err);
}
function zArray(schema) {
	return { parse: (value) => {
		if (!Array.isArray(value)) throw new Error("expected array");
		return value.map((v) => schema.parse(v));
	} };
}

//#endregion
//#region src/monitor.ts
function pollMonitor(config) {
	return async (ctx, emit) => {
		const connectionId = requireConnectionId(ctx);
		let cursor = config.initialCursor ? await config.initialCursor({
			ctx,
			connectionId
		}) : void 0;
		const idleMs = Math.max(0, Math.trunc(config.idleMs ?? 0));
		const errorDelayMs = Math.max(0, Math.trunc(config.errorDelayMs ?? 1e3));
		while (!ctx.signal.aborted) try {
			const page = await config.poll({
				ctx,
				connectionId,
				cursor,
				signal: ctx.signal
			});
			let emitted = 0;
			for (const item of config.items(page)) {
				const result = await config.toRecord({
					ctx,
					connectionId,
					cursor,
					page,
					item
				});
				for (const record of asRecords(result)) {
					emit(record);
					emitted += 1;
				}
				if (config.nextCursor) cursor = await config.nextCursor({
					ctx,
					connectionId,
					cursor,
					page,
					item
				});
			}
			if (config.pageCursor) cursor = await config.pageCursor({
				ctx,
				connectionId,
				cursor,
				page
			});
			if (emitted > 0) ctx.log.debug("emitted %d records", emitted);
			if (idleMs > 0) await sleepWithSignal(ctx.signal, idleMs);
		} catch (err) {
			if (ctx.signal.aborted) break;
			ctx.log.error("poll monitor error: %s", err instanceof Error ? err.message : String(err));
			if (errorDelayMs > 0) await sleepWithSignal(ctx.signal, errorDelayMs);
		}
	};
}
function pollBackfill(config) {
	return async (ctx, args, emit) => {
		const connectionId = requireConnectionId(ctx);
		let cursor = config.initialCursor ? await config.initialCursor({
			ctx,
			connectionId,
			since: args.since
		}) : void 0;
		while (!ctx.signal.aborted) {
			const page = await config.poll({
				ctx,
				connectionId,
				cursor,
				since: args.since,
				signal: ctx.signal
			});
			let itemCount = 0;
			for (const item of config.items(page)) {
				itemCount += 1;
				const result = await config.toRecord({
					ctx,
					connectionId,
					cursor,
					page,
					item
				});
				for (const record of asRecords(result)) emit(record);
				if (config.nextCursor) cursor = await config.nextCursor({
					ctx,
					connectionId,
					cursor,
					page,
					item
				});
			}
			if (config.pageCursor) cursor = await config.pageCursor({
				ctx,
				connectionId,
				cursor,
				page
			});
			if (!(config.hasMore ? await config.hasMore({
				ctx,
				connectionId,
				cursor,
				page,
				itemCount
			}) : itemCount > 0)) return;
		}
	};
}
function asRecords(value) {
	if (!value) return [];
	return Array.isArray(value) ? value : [value];
}
function requireConnectionId(ctx) {
	const connectionId = ctx.connectionId?.trim();
	if (!connectionId) throw new Error("connection id is required");
	return connectionId;
}

//#endregion
//#region src/send.ts
function chunkText(text, limit) {
	if (!text) return [];
	if (limit <= 0 || text.length <= limit) return [text];
	const segments = [];
	const lines = text.split("\n");
	let outside = "";
	let inFence = false;
	let fenceChar = "";
	let fenceLen = 0;
	let openLine = "";
	let body = "";
	const flushOutside = () => {
		if (outside) {
			segments.push({
				kind: "text",
				content: outside
			});
			outside = "";
		}
	};
	const isFenceLine = (lineNoNL) => {
		const trimmed = lineNoNL.replace(/^[ \t]+/u, "");
		if (!trimmed.startsWith("```") && !trimmed.startsWith("~~~")) return null;
		const ch = trimmed[0];
		let n = 0;
		while (n < trimmed.length && trimmed[n] === ch) n++;
		if (n < 3) return null;
		return {
			ch,
			n
		};
	};
	const isFenceClose = (lineNoNL, ch, n) => {
		const trimmed = lineNoNL.replace(/^[ \t]+/u, "");
		if (!trimmed.startsWith(ch.repeat(Math.min(3, n)))) return false;
		let i = 0;
		while (i < trimmed.length && trimmed[i] === ch) i++;
		return i >= n;
	};
	for (let i = 0; i < lines.length; i++) {
		const lineNoNL = lines[i] ?? "";
		const fullLine = i < lines.length - 1 ? `${lineNoNL}\n` : lineNoNL;
		if (!inFence) {
			const fence = isFenceLine(lineNoNL);
			if (fence) {
				flushOutside();
				inFence = true;
				fenceChar = fence.ch;
				fenceLen = fence.n;
				openLine = fullLine;
				body = "";
				continue;
			}
			outside += fullLine;
			continue;
		}
		if (isFenceClose(lineNoNL, fenceChar, fenceLen)) {
			const closeLine = fullLine;
			segments.push({
				kind: "fence",
				open: openLine,
				body,
				close: closeLine
			});
			inFence = false;
			fenceChar = "";
			fenceLen = 0;
			openLine = "";
			body = "";
			continue;
		}
		body += fullLine;
	}
	if (inFence) {
		const closeLine = `${fenceChar.repeat(Math.max(3, fenceLen))}\n`;
		segments.push({
			kind: "fence",
			open: openLine,
			body,
			close: closeLine
		});
	} else flushOutside();
	const chunks = [];
	let current = "";
	const pushCurrent = () => {
		const out = current.replace(/[ ]+$/u, "");
		if (out) chunks.push(out);
		current = "";
	};
	const appendText = (segment) => {
		let remaining = segment;
		while (remaining.length > 0) {
			if (!current) {
				if (remaining.length <= limit) {
					current = remaining;
					return;
				}
				const splitAt = findSplitPoint(remaining, limit);
				const piece = remaining.slice(0, splitAt).replace(/[ ]+$/u, "");
				if (piece) {
					current = piece;
					pushCurrent();
				}
				remaining = remaining.slice(splitAt).replace(/^[ ]+/u, "");
				continue;
			}
			const available = limit - current.length;
			if (available <= 0) {
				pushCurrent();
				continue;
			}
			if (remaining.length <= available) {
				current += remaining;
				return;
			}
			const splitAt = findSplitPoint(remaining, available);
			const piece = remaining.slice(0, splitAt).replace(/[ ]+$/u, "");
			if (piece) {
				current += piece;
				pushCurrent();
			}
			remaining = remaining.slice(splitAt).replace(/^[ ]+/u, "");
		}
	};
	const splitFence = (open, bodyText, close) => {
		const full = `${open}${bodyText}${close}`;
		if (full.length <= limit) return [full];
		const maxBody = limit - (open.length + close.length);
		if (maxBody <= 0) return [full];
		const parts = [];
		const bodyLines = bodyText.split("\n");
		let buf = "";
		const flush = () => {
			if (!buf) return;
			parts.push(buf);
			buf = "";
		};
		for (let i = 0; i < bodyLines.length; i++) {
			const lineNoNL = bodyLines[i] ?? "";
			const line = i < bodyLines.length - 1 ? `${lineNoNL}\n` : lineNoNL;
			if (line.length > maxBody) {
				flush();
				let remaining = line;
				while (remaining.length > 0) {
					const take = Math.min(maxBody, remaining.length);
					parts.push(remaining.slice(0, take));
					remaining = remaining.slice(take);
				}
				continue;
			}
			if (buf.length + line.length > maxBody && buf) flush();
			buf += line;
		}
		flush();
		return parts.filter(Boolean).map((p) => `${open}${p}${close}`);
	};
	for (const seg of segments) {
		if (seg.kind === "text") {
			appendText(seg.content);
			continue;
		}
		if (seg.open.length + seg.body.length + seg.close.length > limit) {
			if (current) pushCurrent();
			const fenceChunks = splitFence(seg.open, seg.body, seg.close);
			for (const c of fenceChunks) {
				const out = c.replace(/[ ]+$/u, "");
				if (out) chunks.push(out);
			}
			continue;
		}
		const fenceText = `${seg.open}${seg.body}${seg.close}`;
		if (!current) {
			current = fenceText;
			continue;
		}
		if (current.length + fenceText.length <= limit) {
			current += fenceText;
			continue;
		}
		pushCurrent();
		current = fenceText;
	}
	if (current) pushCurrent();
	return chunks;
}
function findSplitPoint(text, limit) {
	if (limit >= text.length) return text.length;
	let searchStart = Math.floor(limit * 70 / 100);
	if (searchStart < 0) searchStart = 0;
	let window = text.slice(searchStart, limit);
	let idx = window.lastIndexOf("\n\n");
	if (idx !== -1) return searchStart + idx + 2;
	searchStart = Math.floor(limit * 60 / 100);
	if (searchStart < 0) searchStart = 0;
	window = text.slice(searchStart, limit);
	idx = window.lastIndexOf("\n");
	if (idx !== -1) return searchStart + idx + 1;
	searchStart = Math.floor(limit * 50 / 100);
	if (searchStart < 0) searchStart = 0;
	for (let i = limit - 1; i >= searchStart; i--) {
		const ch = text[i];
		if (ch === "." || ch === "!" || ch === "?") {
			const next = text[i + 1];
			if (next === void 0 || /\s/u.test(next)) return i + 1;
		}
	}
	searchStart = Math.floor(limit * 80 / 100);
	if (searchStart < 0) searchStart = 0;
	window = text.slice(searchStart, limit);
	idx = window.lastIndexOf(" ");
	if (idx !== -1) return searchStart + idx + 1;
	return limit;
}
function sendWithChunking(text, charLimit, sendFn) {
	const totalChars = text ? text.length : 0;
	const chunks = chunkText(text, charLimit);
	if (chunks.length === 0) return Promise.resolve({
		success: false,
		message_ids: [],
		chunks_sent: 0,
		total_chars: totalChars,
		error: {
			type: "content_rejected",
			message: "empty message",
			retry: false
		}
	});
	const messageIds = [];
	return (async () => {
		for (let i = 0; i < chunks.length; i++) {
			const chunk = chunks[i];
			try {
				const id = await sendFn(chunk);
				messageIds.push(id);
			} catch (err) {
				const msg = err instanceof Error ? err.message : String(err);
				return {
					success: false,
					message_ids: messageIds,
					chunks_sent: i,
					total_chars: totalChars,
					error: {
						type: "network",
						message: msg,
						retry: true
					}
				};
			}
		}
		return {
			success: true,
			message_ids: messageIds,
			chunks_sent: chunks.length,
			total_chars: totalChars
		};
	})();
}

//#endregion
//#region src/event.ts
var RecordBuilder = class {
	record;
	constructor(platform, externalRecordID) {
		this.record = {
			operation: "record.ingest",
			routing: {
				platform,
				connection_id: "",
				sender_id: "",
				container_id: "",
				container_kind: "direct"
			},
			payload: {
				external_record_id: externalRecordID,
				timestamp: Date.now(),
				content: "",
				content_type: "text"
			}
		};
	}
	withTimestamp(date) {
		this.record.payload.timestamp = date.getTime();
		return this;
	}
	withTimestampUnixMs(ms) {
		this.record.payload.timestamp = ms;
		return this;
	}
	withContent(content) {
		this.record.payload.content = content;
		return this;
	}
	withContentType(contentType) {
		this.record.payload.content_type = contentType;
		return this;
	}
	withSender(senderID, senderName) {
		this.record.routing.sender_id = senderID;
		if (senderName) this.record.routing.sender_name = senderName;
		return this;
	}
	withReceiver(receiverID, receiverName) {
		this.record.routing.receiver_id = receiverID;
		if (receiverName) this.record.routing.receiver_name = receiverName;
		return this;
	}
	withContainer(containerID, kind, containerName) {
		this.record.routing.container_id = containerID;
		this.record.routing.container_kind = kind;
		if (containerName) this.record.routing.container_name = containerName;
		return this;
	}
	withConnection(connectionID) {
		this.record.routing.connection_id = connectionID;
		return this;
	}
	withThread(threadID, threadName) {
		this.record.routing.thread_id = threadID;
		if (threadName) this.record.routing.thread_name = threadName;
		return this;
	}
	withSpace(spaceID, spaceName) {
		this.record.routing.space_id = spaceID;
		if (spaceName) this.record.routing.space_name = spaceName;
		return this;
	}
	withReplyTo(replyToID) {
		this.record.routing.reply_to_id = replyToID;
		this.record.payload.reply_to_id = replyToID;
		return this;
	}
	withAttachment(attachment) {
		if (!this.record.payload.attachments) this.record.payload.attachments = [];
		this.record.payload.attachments.push(attachment);
		return this;
	}
	withRecipient(recipientID) {
		if (!this.record.payload.recipients) this.record.payload.recipients = [];
		this.record.payload.recipients.push(recipientID);
		return this;
	}
	withMetadata(key, value) {
		if (!this.record.payload.metadata) this.record.payload.metadata = {};
		this.record.payload.metadata[key] = value;
		return this;
	}
	withRoutingMetadata(key, value) {
		if (!this.record.routing.metadata) this.record.routing.metadata = {};
		this.record.routing.metadata[key] = value;
		return this;
	}
	build() {
		return this.record;
	}
};
function newRecord(platform, externalRecordID) {
	return new RecordBuilder(platform, externalRecordID);
}
function messageRecord(options) {
	const builder = newRecord(options.platform, options.externalRecordId).withConnection(options.connectionId).withSender(options.senderId, options.senderName).withContainer(options.containerId, options.containerKind, options.containerName).withContent(options.content).withContentType(options.contentType ?? "text");
	if (options.timestamp instanceof Date) builder.withTimestamp(options.timestamp);
	else if (typeof options.timestamp === "number" && Number.isFinite(options.timestamp)) builder.withTimestampUnixMs(options.timestamp);
	if (options.receiverId) builder.withReceiver(options.receiverId, options.receiverName);
	if (options.spaceId) builder.withSpace(options.spaceId, options.spaceName);
	if (options.threadId) builder.withThread(options.threadId, options.threadName);
	if (options.replyToId) builder.withReplyTo(options.replyToId);
	for (const attachment of options.attachments ?? []) builder.withAttachment(attachment);
	for (const recipient of options.recipients ?? []) builder.withRecipient(recipient);
	for (const [key, value] of Object.entries(options.metadata ?? {})) builder.withMetadata(key, value);
	for (const [key, value] of Object.entries(options.routingMetadata ?? {})) builder.withRoutingMetadata(key, value);
	return builder.build();
}

//#endregion
export { ADAPTER_CONTEXT_ENV_VAR, ADAPTER_STATE_DIR_ENV_VAR, AdapterAccountSchema, AdapterAuthFieldSchema, AdapterAuthManifestSchema, AdapterAuthMethodApiKeySchema, AdapterAuthMethodCustomFlowSchema, AdapterAuthMethodFileUploadSchema, AdapterAuthMethodOAuthSchema, AdapterAuthMethodSchema, AdapterControlEndpointRegistry, AdapterControlEndpointRemoveFrameSchema, AdapterControlEndpointSchema, AdapterControlEndpointUpsertFrameSchema, AdapterControlInputFrameSchema, AdapterControlInvokeCancelFrameSchema, AdapterControlInvokeErrorSchema, AdapterControlInvokeRequestFrameSchema, AdapterControlInvokeResultFrameSchema, AdapterControlOutputFrameSchema, AdapterControlRecordIngestFrameSchema, AdapterControlSession, AdapterHealthSchema, AdapterInboundPayloadSchema, AdapterInboundRecordSchema, AdapterInboundRoutingSchema, AdapterInfoSchema, AdapterMethodCatalogSchema, AdapterMethodContextHintValueSchema, AdapterMethodContextHintsSchema, AdapterMethodOriginSchema, AdapterMethodSchema, AdapterOperationSchema, AdapterSetupResultSchema, AdapterSetupStatusSchema, AdapterStreamStatusSchema, AttachmentSchema, ChannelCapabilitiesSchema, ChannelRefSchema, ContainerKindSchema, ContentTypeSchema, DeliveryErrorSchema, DeliveryErrorTypeSchema, DeliveryResultSchema, DeliveryTargetSchema, RecipientSchema, RecordBuilder, SendRequestSchema, StreamEventSchema, chunkText, createAdapterControlSession, createAdapterLogger, defineAdapter, emitStreamStatus, handleStream, loadAdapterRuntimeContext, loadAdapterStateDir, messageRecord, method, newRecord, parseRetryAfterMs, patchConsoleToStderr, pollBackfill, pollMonitor, readAdapterRuntimeContextFile, readCredential, readReplyToTarget, readThreadTarget, requireAdapterRuntimeContext, requireAdapterStateDir, requireContainerTarget, requireCredential, runAdapter, runAdapter as runAdapterCLI, sendWithChunking, sleepWithSignal, withRetry };
